### PR TITLE
HDDS-6583. Introduce lock strategy pattern implementations based on configuration flags

### PR DESCRIPTION
**Pull request by tanvipenumudy**
## What changes were proposed in this pull request?

This task is intended to introduce lock strategy pattern implementations for the regular `BUCKET_LOCK` and `OBJECT_STORE("OBS")` `KEY_PATH_LOCK`.

The `OBS` `KEY_PATH_LOCK` feature shall be disabled by default using the configuration `ozone.om.enable.key.path.lock` (= false).

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-6583

## How was this patch tested?

Added UTs

### CODE REVIEW

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/lock/OzoneLockStrategy.java [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+package org.apache.hadoop.ozone.om.lock;
+
+import org.apache.hadoop.ozone.om.OMMetadataManager;
+
+import java.io.IOException;
+
+/**
+ * Common interface for lock strategies.
+ */
+public interface OzoneLockStrategy {
```
Comment (kerneltime) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/lock/OzoneLockStrategy.java:27]: Please add detailed Java doc for what this interface guarantees.

Comment (tanvipenumudy) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/lock/OzoneLockStrategy.java:27]: Thanks for the suggestion - will be adding more details to the Java doc!

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/lock/OzoneLockProvider.java [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+package org.apache.hadoop.ozone.om.lock;
+
+import org.apache.hadoop.ozone.om.helpers.BucketLayout;
+
+/**
+ * OzoneLockProvider class returns the appropriate lock strategy pattern
+ * implementation based on the configuration flags passed.
+ */
+public class OzoneLockProvider {
```
Comment (rakeshadr) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/lock/OzoneLockProvider.java:26]: Please add a test case to verify the different strategies and acquire/release locks.

OBS - two locks BUCKET_LOCK and KEY_PATH_LOCK
Regular - only BUCKET_LOCK

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/lock/OzoneLockProvider.java [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+  private boolean enableKeyPathLock;
+  private boolean enableFileSystemPaths;
+
+  public OzoneLockProvider(boolean enableKeyPathLock,
```
Comment (rakeshadr) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/lock/OzoneLockProvider.java:31]: Add another test case where you can check the combinations enableKeyPathLock & enableFileSystemPaths & BucketLayout, then `Assert.assertTrue( lockStrategy instanceof OBSKeyPathLockStrategy)`

Comment (tanvipenumudy) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/lock/OzoneLockProvider.java:31]: Added a UT for `OzoneLockProvider`, thank you for the comment!

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/lock/OzoneLockProvider.java (hunk 2 above)
Comment (tanvipenumudy) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/lock/OzoneLockProvider.java:26]: Added a UT for OzoneLockProvider, thank you for the comment!

File: hadoop-hdds/common/src/main/resources/ozone-default.xml [hunk 4]
```diff
@@ -2819,6 +2819,15 @@
     </description>
   </property>
 
+  <property>
+    <name>ozone.om.enable.key.path.lock</name>
```
Comment (rakeshadr) [inline hadoop-hdds/common/src/main/resources/ozone-default.xml:2823]: Please rename this property `ozone.om.enable.key.path.lock` to` ozone.om.key.path.lock.enabled`.

Also, please rename the corresponding items
-> `OZONE_OM_ENABLE_KEY_PATH_LOCK` to `OZONE_OM_KEY_PATH_LOCK_ENABLED`.

-> `getKeyPathLock() `to `getKeyPathLockEnabled()`

File: hadoop-hdds/common/src/main/resources/ozone-default.xml [hunk 5]
```diff
@@ -2819,6 +2819,15 @@
...
+    <tag>OZONE, OM</tag>
+    <value>false</value>
+    <description>If true, the fine-grained KEY_PATH_LOCK functionality is enabled.
+      If false, it is disabled.
```
Comment (rakeshadr) [inline hadoop-hdds/common/src/main/resources/ozone-default.xml:2827]: You can say, 

Defaults to false.

**Review by rakeshadr — COMMENTED** [review]
Thanks @tanvipenumudy for the patch. Added few comments and also please rebase the PR.

File: hadoop-hdds/common/src/main/resources/ozone-default.xml (hunk 5 above)
Comment (tanvipenumudy) [reply hadoop-hdds/common/src/main/resources/ozone-default.xml]: Thank you, I have incorporated the change!

File: hadoop-hdds/common/src/main/resources/ozone-default.xml (hunk 4 above)
Comment (tanvipenumudy) [reply hadoop-hdds/common/src/main/resources/ozone-default.xml]: I have made the changes, thank you for the comment!

**Review by rakeshadr — APPROVED** [review]
+1 LGTM, Thanks @tanvipenumudy

**Pull request merged**